### ISSUE TITLE
No method matching difference(::VPolygon, ::VPolygon)

### ISSUE THREAD

**Issue by ferrolho** [JuliaReach/LazySets.jl#2346]
How can I subtract a set from another?

I am trying to use https://github.com/JuliaReach/LazySets.jl/blob/48974d76e8d8691bdfdd6d2a24dbbe37015149a8/src/ConcreteOperations/difference.jl#L38
as in the snippet
```julia
board = VPolygon([0.0 1.0 1.0 0.0
                  0.0 0.0 1.0 1.0])

shape₁ = VPolygon([0.2 0.5 0.7
                   0.2 0.9 0.1])

result = board \ shape₁
```
but get the following error:
```
ERROR: MethodError: no method matching difference(::VPolygon{Float64,Array{Float64,1}}, ::VPolygon{Float64,Array{Float64,1}})
Stacktrace:
 [1] \(::VPolygon{Float64,Array{Float64,1}}, ::VPolygon{Float64,Array{Float64,1}}) at /home/henrique/.julia/packages/LazySets/zvXxX/src/ConcreteOperations/difference.jl:38
 [2] top-level scope at REPL[7]:1
```

**Comment by mforets** [comment]
Let me remind that are two non-equivalent notions of set difference:[Minkowski (or Pontryagin) difference](https://en.wikipedia.org/wiki/Minkowski_addition) on one hand, implemented in this library as `minkowski_difference`, and set difference of sets implemented in this libary as `difference(X, Y)` for which `X \ Y` is meant to be an alias and which is defined as the set of points which belong to `X` but not to `Y`. If you are interested in the latter seems like it is not implemented yet unfortunately. I don't remember that we have an open issue for it.

**Comment by mforets** [comment]
If your application is OK with an *approximate* result that's another story, it is not hard to overapproximate (cover) the shape with the set union of boxes, then compute the difference between the board and each element in the shape. In this case, we can estimate the error in the approximation as well and refine the approximation to the desired accuracy.

**Comment by ferrolho** [comment]
Thank you for your replies. I understand the difference, but I am looking for the latter. Picture poking a circular hole through a square piece of paper; that's the kind of operation I am looking for: `result = square \ circle`.

I'm having a look at [Overapproximations](https://juliareach.github.io/LazySets.jl/dev/lib/approximations/#Overapproximations) right now. Is this what you expect to work?

**Comment by mforets** [comment]
> I'm having a look at Overapproximations right now. Is this what you expect to work?

Yes exactly. I tried covering with boxes, the result is shown below and the code is available in [this notebook](https://nbviewer.jupyter.org/github/mforets/escritoire/blob/master/2020/Week35/Difference_VPolygon.ipynb) (the implementation is not very clean due to some missing methods..). I dont understand why there are missing blue rectangles on the edges.

![Screenshot from 2020-08-24 18-49-51](https://user-images.githubusercontent.com/12424594/91099952-a0e3d980-e63a-11ea-9523-214d7ced1817.png)

![Screenshot from 2020-08-24 18-48-51](https://user-images.githubusercontent.com/12424594/91099957-a4776080-e63a-11ea-8f29-38a733d341de.png)

**Comment by ferrolho** [comment]
Thank you for taking the time to prepare a notebook. I will explore this workaround and see where it can get me.

I think I will leave this issue open for now; it can serve as a reminder that _concrete_ difference of sets is not yet implemented.

**Comment by mforets** [comment]
How about using complements? If I didn't make any silly mistake (... it's late here :moon: ), the following formula applies:

![Screenshot from 2020-08-25 23-55-22](https://user-images.githubusercontent.com/12424594/91250107-9e62ac00-e72e-11ea-9c28-420fddd23a0f.png)

Which is easy to implement:

```julia
complement(H::HalfSpace) = HalfSpace(-H.a, -H.b)

function _difference(X, Y)
    out = [intersection(X, complement(H)) for H in constraints_list(Y)]
    return UnionSetArray(out)
end
```
In fact I think the approach works with rather general assumptions on `X` and `Y`, such as `Y` being polyhedral.

The result is shown below, and I've updated the previous notebook [Difference_VPolygon](https://nbviewer.jupyter.org/github/mforets/escritoire/blob/master/2020/Week35/Difference_VPolygon.ipynb) with the new function.

![Screenshot from 2020-08-25 23-55-00](https://user-images.githubusercontent.com/12424594/91250125-a6bae700-e72e-11ea-91c9-7925ed2b6a9c.png)

To hide the black lines delimiting each constraint we can use `lw=0.0`,

![Screenshot from 2020-08-26 00-02-53](https://user-images.githubusercontent.com/12424594/91250516-9a835980-e72f-11ea-87da-0aa6366c4510.png)

**Comment by schillic** [comment]
> How about using complements?

Yes, this idea works as long as `Y` is a convex polyhedron. For more general convex `Y`s you can try to over- or underapproximate `Y` with such a polyhedron first (depending on whether you prefer an under- or overapproximation of the result).

**Comment by schillic** [comment]
I updated my post.

**Comment by ferrolho** [comment]
Thank you @mforets and @schillic for both of your interesting replies.

As suggested above, I have tried using the following snippet:
```julia
complement(H::HalfSpace) = HalfSpace(-H.a, -H.b)

function _difference(X, Y)
    out = [intersection(X, complement(H)) for H in constraints_list(Y)]
    return UnionSetArray(out)
end
```

It worked very well for a single convex shape. Below I have plotted a "before and after" subtracting a triangle from a square.
![Untitled](https://user-images.githubusercontent.com/4166093/91359681-ec3de980-e7ec-11ea-9aa6-74f89de8d796.png)

There are two issues that followed:

1. Trying to get the area results in the following error:
    ```
    ERROR: LoadError: MethodError: no method matching area(::UnionSetArray{Float64,HPolytope{Float64,Array{Float64,1}}})
    Closest candidates are:
    area(::EmptySet{N}) where N at /home/henrique/.julia/packages/LazySets/G2PtZ/src/Sets/EmptySet.jl:377
    area(::LazySet{N}) where N at /home/henrique/.julia/packages/LazySets/G2PtZ/src/Interfaces/LazySet.jl:947
    ```
    Is there a way I can convert the resulting `UnionSetArray` into a `LazySet` for the 2nd `area` method to be dispatched?

2. Indeed, I would like to subtract convex shapes, but not only one single shape. For example, in the figure below, I'd like to subtract all the triangles to the square. However, if I try to call `_difference` (from the code snippet above) for a `UnionSetArray` and a `VPolygon`, I get the following error:
    ```
    MethodError: no method matching UnionSetArray(::Array{UnionSetArray{Float64,S} where S<:LazySet{Float64},1})
    Closest candidates are:
    UnionSetArray(::Array{S,1}) where {N<:Real, S<:LazySet{N}} at /home/henrique/.julia/packages/LazySets/G2PtZ/src/LazyOperations/UnionSetArray.jl:18
    UnionSetArray(::EmptySet{N}, ::EmptySet{N}) where N<:Real at /home/henrique/.julia/packages/LazySets/G2PtZ/src/Utils/macros.jl:45
    UnionSetArray(::Universe{#283#N}, ::EmptySet{#283#N}) where #283#N<:Real at /home/henrique/.julia/packages/LazySets/G2PtZ/src/Utils/macros.jl:258
    ...
    Stacktrace:
    [1] _difference(::UnionSetArray{Float64,HPolytope{Float64,Array{Float64,1}}}, ::VPolygon{Float64,Array{Float64,1}}) at /home/henrique/tests/test_lazy_sets.jl:31
    [2] top-level scope at /home/henrique/tests/test_lazy_sets.jl:35
    [3] include(::String) at ./client.jl:439
    [4] top-level scope at REPL[2]:1
    in expression starting at /home/henrique/tests/test_lazy_sets.jl:35
    ```

**Comment by schillic** [comment]
1. Computing the area of a union is tricky if the components overlay. I guess you would essentially have to do the work twice for that. If you are ultimately only interested in the area of the difference, I think it is better to instead compute the area of the intersection because that will be convex:

        area(board \ triangle) = area(board) - area(board ∩ triangle)

    (The philosophy to realize this idea in `LazySets` would be to define a lazy `Difference` type and then define `area(::Difference)` as given above.)

2. With multiple objects, if you go sequentially, after the first difference you are left with a non-convex set, and that complicates things considerably.
I think if you are only interested in the area, you can again use a similar idea, but need to take care of overlays. Here is the code for two triangles.

        area(board \ (t1 ∪ t2)) = area(board) - area(board ∩ t1) - area(board ∩ t2) + area(board ∩ t1 ∩ t2)

    (The generalization to and arbitrary number of sets alternates `-` and `+`:

        area(board) - ∑_i area(board ∩ ti) + ∑_i≠j area(board ∩ ti ∩ tj) - ∑_i≠j≠k area(board ∩ ti ∩ tj ∩ tk) + ...

    An efficient implementation would need to track intersections.)
